app/grandchallenge/notifications/views.py

```diff
@@ -1,27 +1,113 @@
-from django.contrib.auth.mixins import LoginRequiredMixin
-from django.utils.timezone import now
-from django.views.generic import TemplateView
+from actstream.models import Follow
+from django.contrib import messages
+from django.contrib.messages.views import SuccessMessageMixin
+from django.http import (
+    HttpResponseNotFound,
+    HttpResponseRedirect,
+)
+from django.template.defaultfilters import pluralize
+from django.views.generic import CreateView, DeleteView, ListView
+from guardian.mixins import (
+    LoginRequiredMixin,
+    PermissionListMixin,
+    PermissionRequiredMixin as ObjectPermissionRequiredMixin,
+)
+from guardian.shortcuts import get_objects_for_user
 
-from grandchallenge.profiles.models import UserProfile
+from grandchallenge.notifications.models import Notification
+from grandchallenge.notifications.utils import (
+    prefetch_generic_foreign_key_objects,
+    prefetch_nested_generic_foreign_key_objects,
+)
+from grandchallenge.subdomains.utils import reverse
 
 
-class NotificationList(LoginRequiredMixin, TemplateView):
-    template_name = "notifications/notification_list.html"
+class NotificationList(LoginRequiredMixin, PermissionListMixin, ListView):
+    model = Notification
+    permission_required = "view_notification"
+    paginate_by = 50
 
-    def get_context_data(self, **kwargs):
-        context = super().get_context_data(**kwargs)
+    def get_queryset(self):
+        return prefetch_nested_generic_foreign_key_objects(
+            super().get_queryset().order_by("-created")
+        )
 
-        profile = self.request.user.user_profile
+    def post(self, request, *args, **kwargs):
+        if "delete" in request.POST:
+            action = "delete"
+        elif "mark_read" in request.POST:
+            action = "mark_read"
+        elif "mark_unread" in request.POST:
+            action = "mark_unread"
+
+        selected_notifications = request.POST.getlist("checkbox")
+        notifications = get_objects_for_user(
+            request.user,
+            ["delete_notification", "change_notification"],
```

Comment:
Split these permissions out so that you're only looking for one, eg, where you're setting `action = "mark_read"` also set `required_permission = "change_notification"`, and then only look for the required permission. This will help when refactoring.